### PR TITLE
[IMP] web, *: clean up notification API

### PR DESCRIPTION
**Pull request by seb-odoo**
task-2476867

### CODE REVIEW

File: addons/web/static/src/js/core/misc.js
Comment (xmo-odoo) [inline addons/web/static/src/js/core/misc.js:194]: If I understand correctly `isMessageHtml` only affects the message, the `title` is now always escaped. This means existing explicit title escapings (as with this call) must be removed or we'll have double-escaped titles.

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (xmo-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js:2436]: The translation seens completely broken here. Even if babel can extract the source string, `_t` is going to receive the string post-interpolation.

File: addons/web_editor/static/tests/field_html_tests.js
Comment (xmo-odoo) [inline addons/web_editor/static/tests/field_html_tests.js:151]: Probably requires `isMessageHtml`?

File: addons/web/static/src/js/core/service_mixins.js
Comment (xmo-odoo) [inline addons/web/static/src/js/core/service_mixins.js:247]: maybe an imperative `messageIsHtml`? The `isX` form is more of a query e.g. checking for a widget's state or somesuch.

File: addons/web/static/src/js/core/service_mixins.js
Comment (xmo-odoo) [inline addons/web/static/src/js/core/service_mixins.js:241]: That's ancillary to this change, but maybe we should check if there's a title or a message? Seems useless to show a notification with neither.

File: addons/web/static/src/js/core/misc.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js:194]: I'm still considering options here. From what I've seen so far, the rare case were we use html is to format a list, so I might as well allow a list as param instead.

But yes either way I would have to go over all existing occurrences doing an escape and removing the escape.

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (seb-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js:2436]: I guess the issue is already present before the PR then? I'll double check and suggest fix in the appropriate version.

File: addons/web_editor/static/tests/field_html_tests.js
Comment (seb-odoo) [reply addons/web_editor/static/tests/field_html_tests.js:151]: Good catch, either that or again an example where it's just a list.

File: addons/web/static/src/js/core/service_mixins.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/service_mixins.js:247]: Well for me the boolean is the answer to a question (Is message html?) which usually comes with the verb at the start.
Also helps to detect the type at first sight. Quickly reading something that starts with "message..." I would first think it contains a message (or something else, but I'd need to read everything more carefully).
Anyway as per my first comment, I might remove the possibility to use html all together if all it is ever used for is to pass a list.

File: addons/web/static/src/js/core/service_mixins.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/service_mixins.js:241]: It's a programming error but if we can help developers detect the mistake why not.

File: addons/web/static/src/js/core/misc.js
Comment (xmo-odoo) [reply addons/web/static/src/js/core/misc.js:194]: > Anyway as per my first comment, I might remove the possibility to use html all together if all it is ever used for is to pass a list.

That's not the case e.g. some of the notifications bounced from the server (via the `display_notification` client action) contain links for instance. I know that @JKE-be also mentioned client-side notifications with various formatting, though I don't remember where so I don't have a link to provide.

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (xmo-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js:2436]: Yes, I just saw that because the PR touches it but it should be broken regardless.

File: addons/web/static/src/js/core/service_mixins.js
Comment (xmo-odoo) [reply addons/web/static/src/js/core/service_mixins.js:247]: > Well for me the boolean is the answer to a question (Is message html?) which usually comes with the verb at the start.

Yes, but here it's not a question, the caller is telling the function that the message is HTML.

File: addons/web/static/src/js/core/misc.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js:194]: Fortunately the links are passed as a list from `display_notification` and not as html. So at the end it's just a list again, though indeed this would require a different list format to add the links compared to the more generic list.

Comment (xmo-odoo) [reply addons/web/static/src/js/core/misc.js:194]: > Fortunately the links are passed as a list from `display_notification` and not as html. So at the end it's just a list again, though indeed this would require a different list format to add the links compared to the more generic list.

It's more complicated than that. The links are passed as a special list of items so that they can be properly escaped and merged in the message independently. The message is not actually a list of anything, it's text with links embedded.

Comment (JKE-be) [reply addons/web/static/src/js/core/misc.js:194]: here an example that contains html in the notification

https://github.com/odoo/odoo/blob/14.0/addons/calendar/static/src/xml/notification_calendar.xml#L12

Not sure how you plan to make it easy to send a notification with link, button, ... with a list :/:
"You need to <b>configure</b> your api key here: <a>click here</a> or use another map provider.

You can choose <a>here</a> you default map provider"

Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js:194]: It doesn't change the point that it's possible to forward the links to the notification itself and make it part of its options. It's just moving the logic one level down, so that any html handling is done in the template with t-att/t-esc and nowhere else.

If there is a title display it, if there is a message display it, if there are links display them after the message, if there are list items display them below the rest with an actual list tag. This can be expanded to other needs too.

Advantage: no more html issues, consistent look for all notifications, easier code to maintain
Disadvantage: no easy customization (but do we really need it?)

I would also like to note we have an option to convert these notifications to native OS notifications, which don't support html anyway. Accepting only raw data and formatting them according to the destination would make handling this much easier.


That being said, thanks for the example in `calendar`. If we really want to support such cases, then we would definitely need to support a formatting language, maybe markdown as suggested by @mart-e. Or just keep the concept of `isMessageHtml` introduced in this PR, in current state html is still possible. As long as it is not html by default it's already much better.

Comment (JKE-be) [reply addons/web/static/src/js/core/misc.js:194]: imo it is not because OS cannot show a bold word in a notification that it cannot be used to make it more readable in webclient. And it is frequent that web_editor show notification with link to backend setting to configure something. 
So yes, we should probably keep the html feature, even if opt-out by default, at least keep to support it when needed.

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (mart-e) [reply addons/web/static/src/js/views/basic/basic_model.js:2436]: I think we were talking about forbidding the ```_t(`...)``` with @d-fence

Comment (seb-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js:2436]: Just ``` _t(`...`) ``` can be useful for example if there are both `"` and `'` in the string it's convenient to have another delimiter to avoid having to escape anything. It's the use of `${}` inside it that is an issue I suppose?

File: addons/web/static/src/js/core/misc.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js:194]: Just an idea but another alternative to allow HTML in a safer way would be to accept a template name + parameters (or OWL equivalent component/props) to ensure the injected HTML comes from a template and not from manually built HTML.

Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js:194]: This is also very similar to the concept of slot in OWL https://github.com/odoo/owl/blob/master/doc/reference/slots.md that is used for example in Dialog and Popover, where Notification in that case would just be a "container" positioned appropriately and would receive its content from its parent template.

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (d-fence) [reply addons/web/static/src/js/views/basic/basic_model.js:2436]: Yes, there will be a linter. Already working in that [branch](https://github.com/odoo-dev/odoo/tree/master-lint-template-strings-moc). No PR yet

File: addons/web/static/src/js/core/misc.js
Comment (xmo-odoo) [reply addons/web/static/src/js/core/misc.js:194]: > Just an idea but another alternative to allow HTML in a safer way would be to accept a template name + parameters (or OWL equivalent component/props) to ensure the injected HTML comes from a template and not from manually built HTML.

If the API accepts a literal template (not a template name but the actual contents of the template), it doesn't fix anything because you can trivially create an unsafe template. If the API does not accept a literal template, it can't do the job.

I was thinking of markdown for these situations, that allows for restricted customisations, there are libraries which allow enabling and disabling features as necessary, and it's trivial to disable the "HTML passthrough" feature. There are a few questions there (buttons, styling) but it does provide a good basis. And there are extensions available in existing markdown libraries e.g. https://www.npmjs.com/package/markdown-it-attrs

File: addons/web/static/src/js/views/basic/basic_model.js
Comment (seb-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: This was changed, I rebased on it.

File: addons/web/static/src/js/core/service_mixins.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/service_mixins.js]: Ok I don't mind either way, I renamed it to your suggestion for now.

File: addons/web/static/src/js/core/service_mixins.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/service_mixins.js:241]: I tried to prevent it but as per test "Display a custom notification" we apparently want to support case where the notification content is fully custom and so none of those should be set. We could probably design it better around this, but as all I really want to do here is to make it safe by default I'll return to no check here.

File: addons/web/static/src/js/chrome/action_manager.js
Comment (seb-odoo) [inline addons/web/static/src/js/chrome/action_manager.js:536]: @mart-e translations are handled correctly for string that are split this way?

File: addons/web/static/src/js/core/misc.js
Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js]: I've removed the double escaped title here, as per the first suggestion. I haven't spotted another title that was already escaped, though sometimes it is a bit complex to figure out where the information is coming from...

Comment (seb-odoo) [reply addons/web/static/src/js/core/misc.js]: As for the rest of the discussion, I propose that we keep the proposition of current PR, which is making it safe by default, but still customizable with HTML by passing a special flag (that you can mark in the ci/security).

Transforming them markdown is out of scope for me, while I can see the added value, this should be another task.

**Review by ged-odoo — APPROVED** [review]

File: addons/stock/static/src/js/inventory_singleton_list_controller.js
Comment (mart-e) [inline addons/stock/static/src/js/inventory_singleton_list_controller.js:55]: "You have tried to create a record which already exists"

File: addons/web/static/src/js/chrome/action_manager.js
Comment (mart-e) [reply addons/web/static/src/js/chrome/action_manager.js:536]: Looks like babel handles it fine
https://github.com/odoo/odoo/blob/master/addons/web/i18n/web.pot#L391

File: addons/stock/static/src/js/inventory_singleton_list_controller.js
Comment (seb-odoo) [reply addons/stock/static/src/js/inventory_singleton_list_controller.js:55]: Ok, I'll take the opportunity to change it.

File: addons/web/static/src/js/chrome/action_manager.js
Comment (seb-odoo) [reply addons/web/static/src/js/chrome/action_manager.js:536]: Ok good to know, thank you for checking.

**Comment by seb-odoo** [comment]
Thank you for checking @ged-odoo, and all other reviewers so far of course.

If there is no other objection for the current diff, I'd like to ask @odoo/rd-security to double check what has been done (in particular the template change itself and usage of `messageIsHtml`) and validate the CI accordingly (and also merge the PR I guess).

Reasons:
- no `t-raw` is actually introduced, 2 have been removed and one has been made conditional
- the said condition is on `messageIsHtml`, and all places where it is set should have been escaped

Suggestion: add `messageIsHtml` to security check after the PR is merged.

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

Already added `messageIsHtml` to the security check, no reason not to.

One thing I'm wondering: were the server-side HTML notification things from website & stock tested? I don't really remember where they were but I do remember they were quite insistent.

**Comment by seb-odoo** [comment]
@xmo-odoo I double-checked once more all of:
- `displayNotification` (`ServiceProviderMixin`)
- `notify` (`NotificationService` trigger up or direct call)
- `sendNotification` (indirection from bus for OS notif)

The one from the server are `display_notification` (sent on bus from server) but you already escaped all of them in 6f4f8f0265d3f7c1d7a2561c1b8a3697e1a8ecc5.

I'm not aware of other means to display this kind of notification, so if there is, I didn't check them.
The calendar example that was linked earlier looks like an override of the template to add buttons after it, so I don't think it's impacted by the change.

I found a few more that I now changed to also escape + use html.
These new flag `messageIsHtml` (since your last comment) are in:
- `addons/mail/static/src/widgets/discuss_invite_partner_dialog/discuss_invite_partner_dialog.js`
- `addons/web_editor/static/src/js/editor/snippets.editor.js`
- `addons/website/static/src/js/content/website_root.js`
- `addons/website/static/src/js/editor/snippets.options.js`
- `addons/website_event/static/src/js/register_toaster_widget.js`
- `addons/website_slides_survey/static/src/js/slides_certification_upload_toast.js`

I also removed plenty of now obsolete escapes in `addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js` and a few other files.


For the sake of transparency, I didn't actually test each notification changed in this PR, as this would be way too tedious.

I double-checked the various calls more times than I should probably have at this point, but that still doesn't mean I didn't miss any of them.
Just due to the sheer amount, it would be expected that I missed some here or there, and worse case scenario they will now display html tags as text (which is obviously better than the other way around). If it indeed happens that some have been missed they should quickly get reported and fixed.

Note that the more this PR is delayed, the higher the chance some new notifications are introduced at rebase (and therefore not caught by my previous checks).

**Comment by ged-odoo** [comment]
@seb-odoo any plan to merge this soon?

**Comment by seb-odoo** [comment]
@ged-odoo for me it was ok, but I made changes to sensitive code (as documented in my last message) that were not yet checked by the @odoo/rd-security team.

But now the PR is out of date so I would need to rebase and double check everything again.

**Comment by seb-odoo** [comment]
Rebased, fixed conflicts, adapted newly introduced calls, moved doc to new repo (https://github.com/odoo/documentation/pull/999)

@odoo/rd-security could you please make the final check, especially around where `messageIsHtml` is used and its corresponding escape?

File: addons/web/static/src/js/core/misc.js
Comment (mart-e) [inline addons/web/static/src/js/core/misc.js:190]: maybe add a comment that the title is already escaped in the template

**Comment by mart-e** [comment]
All the others `messageIsHtml` looks controlled :+1:

**Comment by seb-odoo** [comment]
@mart-e Thank you for the check, I'm not sure if the extra comment was really useful but I doesn't hurt to have it so I added it.

I guess it's time to move forward now.

@robodoo r+

**Pull request closed without merging**
